### PR TITLE
run moto tests in the ci

### PR DESCRIPTION
**Pull request by steffyP**
This PR enables the initial moto integration test run against LocalStack.

## Background Information
The goal is to collect additional metrics for our LocalStack coverage page, and identify all operation-calls that rely on a moto-fallback. Those ones should be covered by moto tests, and the test is expected to succeed.

## Implementation
We use a `conftest.py` with some patches to run the moto-integration test suite against LocalStack.

A LocalStack Extension is used to retrieve the recorded metrics from LocalStack, and store these raw metrics data into a csv - similar as we already do it for the community and pro integration tests.

By default we run tests that match a certain pattern (e.g. contain the name `test_{service}`), for all services that are supported by LocalStack.

EKS and one specific rds-test case are currently excluded, because those take too many resources and causes CI to crash.

The pipeline is expected to fail, because not all tests will run successfully. We only collect the data from tests that run successfully and store the result as an artifact. The artifact will then be retrieved by docs to update the coverage overview.

## Open Questions/Discussion Points
* Currently using the dev-mode for the extensions
  * Could extract the extension into another repo, if we want to re-use it for e.g. terraform test collection
* Is there a better way to use the `login` functionality in CI? Can we create a dedicated user for our test?

### CODE REVIEW

File: collect-raw-metric-data-extension/collect_raw_metric_data_extension/extension.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+    def update_response_handlers(self, handlers: aws.CompositeResponseHandler):
+        pass
+
+
+def _create_simple_dict(metric):
+    """ creates a simple dict represenation of the metric
+    currently only considering metrics we are interested in/use for docs coverage """
+    return {"service": metric.service,
+            "operation": metric.operation,
```
Comment (thrau) [inline collect-raw-metric-data-extension/collect_raw_metric_data_extension/extension.py]: nit: could we run black code formatting?

**Review by thrau — APPROVED** [review]
I think this is a pretty good solution for a first iteration! Of course I love the extension approach :-)

I would maybe have put the start/stop logic for localstack in the github action rather than using `os.system`, but I guess it's fine for now :-) In fact, this is a good case for an API that allows us to start/stop localstack programmatically using python. we have something like that here for instance https://github.com/localstack/cloudpods-server/blob/6c2af4b7cb37a1fd3a0765e220a342b010690ad2/cloudpods_server/backend/server/container.py#L30-L42. but totally out of scope of this PR.

great to see we have a first working version! :partying_face:

**Comment by steffyP** [comment]
> I think this is a pretty good solution for a first iteration! Of course I love the extension approach :-)

@thrau: Regarding the "discussion points" I outlined in the description: 
What do you think about the `login` for using the extensions, e.g. what would be a good pattern to handle this in CI? Ideally, there would be a token that we could use, but I think we do not support this yet. Should I just create a new dedicated account for this purpose?

File: collect-raw-metric-data-extension/collect_raw_metric_data_extension/extension.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+from localstack.extensions.api import Extension, aws, http
+from localstack.http import Request
+
+
+class MyExtension(Extension):
+    name = "collect-raw-metric-data-extension"
+
+    def on_extension_load(self):
+        # enable the metric collection by setting the env
+        os.environ["LOCALSTACK_INTERNAL_TEST_COLLECT_METRIC"] = "1"
```
Comment (dominikschubert) [inline collect-raw-metric-data-extension/collect_raw_metric_data_extension/extension.py:14]: Would love if we maybe had a proper modifiable config at some point that is accessible here in a typed way :thinking: 

But that's a task for future LocalStack :sunglasses:

File: conftest.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+os.environ["MOTO_CALL_RESET_API"] = "false"
+settings.TEST_SERVER_MODE = True
+
+
+@pytest.fixture(scope="session", autouse=True)
+def default_localstack_client_config_fixture() -> Iterator[None]:
+    def localstack_client(*args, org=boto3.client, **kwargs):
+        # kwargs["endpoint_url"] = 'http://localhost:4566' # TODO could also use fixture to override endpoint_url
+        config: Config = kwargs.get("config")
+        max_retry_config = Config(retries={"max_attempts": 1, "total_max_attempts": 2})
```
Comment (dominikschubert) [inline conftest.py:33]: Did you run into cases where the retry made it more stable or did you add this just-in-case? :thinking:

File: conftest.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+            metric_json = json.loads(metric_response.content.decode("utf-8"))
+
+            with open(FNAME_RAW_DATA_CSV, "a") as fd:
+                writer = csv.writer(fd)
+                for m in metric_json.get("metrics"):
+                    m["node_id"] = node_id
+                    writer.writerow(m.values())
+        except json.JSONDecodeError:
+            print("could not decode metrics")
+            pass
```
Comment (dominikschubert) [inline conftest.py]:
```suggestion
```

**Review by dominikschubert — APPROVED** [review]
Looks good. 

Thomas already pointed out the concern regarding startup/shutdown, otherwise the logic seems sound :+1:

File: conftest.py (hunk 3 above)
Comment (steffyP) [reply conftest.py:33]: Yes, actually. The first test runs took way longer, due to a lot of retries for requests that were failing on LocalStack.

**Comment by steffyP** [comment]
Running the tests again, after the approval, revealed new issues. It seems like too many resources were not properly cleaned up (`The runner has received a shutdown signal.`). 

* To mitigate this issue, the fixture that starts/shutsdown localstack is set to `module` now. I also tried to use `package` it had the same effect as using `session`. 
* Additionally, another test was hanging, because it did not receive any response. I therefore added the `pytest-timeout` lib, and set a timeout for test cases.
* Also added the same unittest report action that the terraform tests use to get a quick overview of the test results.

/cc @thrau @dominikschubert

**Review by dominikschubert — APPROVED** [review]
Looked over the recent changes. All LGTM :+1:

**Pull request merged**